src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,438 @@
+            Err(e) => {
+                net_debug!("DHCP error parsing pkt from {}: {:?}", src_ip, e);
+                return None;
+            }
+        };
+        let mac = iface.ethernet_addr();
+        if dhcp_repr.client_hardware_address != mac { return None }
+        if dhcp_repr.transaction_id != self.transaction_id { return None }
+        let server_identifier = match dhcp_repr.server_identifier {
+            Some(server_identifier) => server_identifier,
+            None => return None,
+        };
+        net_debug!("DHCP recv {:?} from {} ({})", dhcp_repr.message_type, src_ip, server_identifier);
+
+        let config = if (dhcp_repr.message_type == DhcpMessageType::Offer ||
+                         dhcp_repr.message_type == DhcpMessageType::Ack) &&
+                      dhcp_repr.your_ip != Ipv4Address::UNSPECIFIED {
+               let address = dhcp_repr.subnet_mask
+                   .and_then(|mask| IpAddress::Ipv4(mask).to_prefix_len())
+                   .map(|prefix_len| Ipv4Cidr::new(dhcp_repr.your_ip, prefix_len));
+               let router = dhcp_repr.router;
+               let dns_servers = dhcp_repr.dns_servers
+                   .unwrap_or([None; 3]);
+               Some(Config { address, router, dns_servers })
+        } else {
+            None
+        };
+
+        match self.state {
+            ClientState::Discovering
+                if dhcp_repr.message_type == DhcpMessageType::Offer =>
+            {
+                self.next_egress = now;
+                let r_state = RequestState {
+                    retry: 0,
+                    endpoint_ip: *src_ip,
+                    server_identifier,
+                };
+                Some(ClientState::Requesting(r_state))
+            }
+            ClientState::Requesting(ref r_state)
+                if dhcp_repr.message_type == DhcpMessageType::Ack &&
+                   server_identifier == r_state.server_identifier =>
+            {
+                self.next_egress = now + Duration::from_secs(RENEW_INTERVAL);
+                let p_state = RenewState {
+                    retry: 0,
+                    endpoint_ip: *src_ip,
+                    server_identifier,
+                };
+                Some(ClientState::Renew(p_state))
+            }
+            ClientState::Renew(ref mut p_state)
+                if dhcp_repr.message_type == DhcpMessageType::Ack &&
+                server_identifier == p_state.server_identifier =>
+            {
+                self.next_egress = now + Duration::from_secs(RENEW_INTERVAL);
+                p_state.retry = 0;
+                None
+            }
+            _ => None
+        }.map(|new_state| self.state = new_state);
+
+        config
+    }
+
+    fn egress<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, raw_socket: &mut RawSocket, checksum_caps: &ChecksumCapabilities, now: Instant) -> Result<Option<Config>> {
+        println!("DHCP egress");
```

Comment:
can you remove these debugging statements? They are hard to use on `no_std` ;)

Reply:
Ooops, thanks for the find!

Nice talk on saturday, btw.